### ISSUE TITLE
[2.x] InfluxDB bucket stops reading+writing every couple of days

### ISSUE THREAD

**Issue by nward** [influxdata/influxdb#23956]
__Steps to reproduce:__
List the minimal actions needed to reproduce the behavior.

1. Run influxdb2
2. Insert metrics (with telegraf)
3. Wait for some time

__Expected behavior:__
Things keep working

__Actual behavior:__
InfluxDB2 main index ("telegraf") stops reading/writing data.
Other indexes work fine - including one which is 5m aggregates of the telegraf raw index. (obviously this does not get any new data)

We have had this in the past randomly, but in the last few weeks has happened every few days.
In the past it seemed to happen at 00:00UTC when influx did some internal DB maintenance - but now happens at random times.

__Environment info:__

* System info: Linux 3.10.0-1160.66.1.el7.x86_64 x86_64
* InfluxDB version: InfluxDB v2.3.0+SNAPSHOT.090f681737 (git: 090f681737) build_date: 2022-06-16T19:33:50Z
* Other relevant environment details: CentOS 7 on vmware - lots of spare IO, CPU, memory.

Our database is 170GB, mostly metrics inserted every 60s, some every 600s.
storage_writer_ok_points is around 2.5k/s for 7mins, then ~25k/s for 3mins for the every-600s burst.

VM has 32G RAM, 28G of which is in buffers/cache.
4 cores, and typically sits at around 90% idle.
~ 24IOPS, 8MiB/s

__Config:__

```
bolt-path = "/var/lib/influxdb/influxd.bolt"
engine-path = "/var/lib/influxdb/engine"
flux-log-enabled = "true"
```

We have enabled flux-log to see if specific queries are causing this - but it doesn't seem to be.

__Logs:__
Include snippet of errors in log.

__Performance:__

I captured a 10s pprof which I will attach.

I also have a core dump, and a 60s dump of debug/pprof/trace (though not sure if this has sensitive info but can share privately - the core dump certainly will)

**Comment by nward** [comment]
[all.tar.gz](https://github.com/influxdata/influxdb/files/10112280/all.tar.gz)

**Comment by jeffreyssmith2nd** [comment]
Hey @nward, I've looked through the profiles attached and haven't seen anything that immediately highlights the issue. If you are able to share more data as mentioned in the issue, I can provide you with a private sftp link. If that works for you, please email me at jsmith @ influxdata.com and I'll get those details to you. Thanks!

**Comment by nward** [comment]
> Hey @nward, I've looked through the profiles attached and haven't seen anything that immediately highlights the issue. If you are able to share more data as mentioned in the issue, I can provide you with a private sftp link. If that works for you, please email me at jsmith @ influxdata.com and I'll get those details to you. Thanks!

I've emailed you now - cheers!

**Comment by zekena2** [comment]
Hi @nward, Any update on this problem? I'm experiencing it for months already.

**Comment by nward** [comment]
> Hi @nward, Any update on this problem? I'm experiencing it for months already.

Hi @zekena2 - yes @jeffreyssmith2nd sent me details to upload my core dumps which I'll do soon. Sorry, just getting back into work here after the holidays :)

**Comment by zekena2** [comment]
hi @jeffreyssmith2nd Is there any updates on this? it's still happening in peak hours.

**Comment by nward** [comment]
Thanks for the reminder @zekena2. I've uploaded one of our core files now @jeffreyssmith2nd, to the SFTP you provided. Let me know if there's more you need.

We have worked around this issue with a script that restarts influxdb if there are not writes for a few minutes, but we can probably update that script to generate new core files if that would be useful.

**Comment by jeffreyssmith2nd** [comment]
Hey thanks for uploading that @nward, do you know by chance if you created the core dump with `GOTRACEBACK=crash` set?

**Comment by nward** [comment]
I don't believe so - we are running influxdb from the RPMs and aren't modifying any environment variables.

Note that the process is not crashing - it is simply stoping working. We create a core file before restarting it. Given that, I don't think `GOTRACEBACK=crash` would have any effect.

**Comment by jeffreyssmith2nd** [comment]
Good catch, that would only apply for a crash. The core file appears to be corrupted (getting a lot of `unreadable error derefing *G error while reading spliced memory at ...`), can you check that the ulimit is set properly before you generate it?

A couple of other thoughts:
- Do you have compactions running regularly and succeeding?
- Do you have the ability to test on a later version and see if the issue is still present?
- You mentioned a 60s trace file, if you can throw that on the ftp server too that may provide some insight.

**Comment by zekena2** [comment]
In my case file limits aren't an issue but when this situation happen the number of open files and unix sockets as well as memeroy keeps growing without stopping until restarting (This is actually how I know there's a problem and my script restart influxdb afterward). I upgraded to the latest version and ironically it happened the same night I did the upgrade. I'll try to monitor the compactions and let you know.

**Comment by nward** [comment]
Hi @jeffreyssmith2nd

> Good catch, that would only apply for a crash. The core file appears to be corrupted (getting a lot of `unreadable error derefing *G error while reading spliced memory at ...`), can you check that the ulimit is set properly before you generate it?

Yes, ulimit is set correctly. It could be that the disk ran out of space though, I recall one of the engineers ran in to that problem at some point, maybe it was that file I uploaded. I have a bunch of core files, I'll upload the lot.

> * Do you have compactions running regularly and succeeding?

Yes we have compactions running and.. well they're not failing, though I don't see an explicit "success" type message. I see beginning compaction, TSM compaction start, TSM compaction end. We see this.. ~1375 times per day.
We have not configured anything special here - it is the default config from the RPM.

> * Do you have the ability to test on a later version and see if the issue is still present?

Yes we can test on a later version. We had not upgraded around the time of this issue as from memory there was a bunch of open 2.5 performance issues or something like that.

> * You mentioned a 60s trace file, if you can throw that on the ftp server too that may provide some insight.

I'm not sure if I still have the trace file, I'll see what I can dig up.

**Comment by nward** [comment]
OK - I have uploaded 4 more core dumps from various times (I am not sure if they are complete I am sorry).
I have also uploaded the trace file that I captured when this issue was happening.

We have an aggregation query running every 5 mins storing data in to another bucket, and that appears to still run, though of course without the data from the main bucket that gets stuck - only one bucket is impacted by this issue, so far as I can tell.

**Comment by jeffreyssmith2nd** [comment]
After looking at the trace, there is something hammering the `HasTagValue` call which ultimately stems from a `MeasurementNamesByPredicate` call. This implies to me there's some a query running that is filtering by measurement/tag value and apparently quite frequently (300,000 calls to `HasTagValue` in that 1min trace).

You mentioned earlier that you have flux logging enabled, can you share the queries that are happening within a few minutes of the issue?

Also, can you provide any error logs from both influxdb and Telegraf when the lockup happens?

**Comment by zekena2** [comment]
It happened half an hour ago and all the logs are info logs with msg between "Executing query" and " executing new query" which some of the quries are just for alerting and some are actual users watching a dashboard.

Telegraf is filled with context deadline exceeded (Client.Timeout exceeded while awaiting headers) during the hanging
Influxdb didn't have any error logs.

I can send you logs by email if you want. They contain the queries as well.

**Comment by nward** [comment]
> After looking at the trace, there is something hammering the `HasTagValue` call which ultimately stems from a `MeasurementNamesByPredicate` call. This implies to me there's some a query running that is filtering by measurement/tag value and apparently quite frequently (300,000 calls to `HasTagValue` in that 1min trace).

That could well be the case. I don't know the internals of InfluxDB to know exactly what those methods do, or why they are called - but we have a lot of series and filter them with tags.

Note that only the primary data bucket is impacted, and only for writes.
Reads are OK, and a scheduled task which runs every 10 mins reading from the primary bucket and writing to an aggregated data bucket is fine - it just doesn't write much data after the primary bucket stops writes.

> You mentioned earlier that you have flux logging enabled, can you share the queries that are happening within a few minutes of the issue?
> 
> Also, can you provide any error logs from both influxdb and Telegraf when the lockup happens?

Influxdb doesn't have any error messages from memory.. I'll have a look We have flux logging enabled, so I'll email you the full influx logs leading up to, and just after the issue time. Not secret, but, not public enough to post somewhere that Google will index :-)


In telegraf we get "did not complete within its flush interval" for influxdb_v2 output from time to time normally (every few minutes). At the time of the issue, that picks up to maybe 6-7 times per minute.

At the time of the issue, in telegraf we also get:
```
Error writing to outputs.influxdb_v2: Post "http://xxxx:8086/api/v2/write?bucket=telegraf&org=xxxx": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
```

This happens maybe once every second or two.

**Comment by zekena2** [comment]
Any updates on this issue?

**Comment by jeffreyssmith2nd** [comment]
After reviewing the logs, it looks like both reading and compactions stop on the bucket that is no longer allowing writes. My assumption is something has a lock on that bucket, usually this is caused by deletes. @zekena2 and @nward, are you doing deletes on the buckets that stop receiving writes? Also, @nward do you see the open file/socket issue that @zekena2 mentioned seeing?

**Comment by zekena2** [comment]
We don't have any deletes it's only writes and reads except of course the bucket retention which is 30d.

**Comment by zekena2** [comment]
I use this to get the sockets number `sar -n SOCK 1 1 | tail -n 1 | awk '{ print $2 }'` Which probably exists under /proc/net/sockets in the first line 
`sockets: used 189`

More info [here](https://serverfault.com/questions/400908/sar-what-does-totsck-stand-for)

**Comment by nward** [comment]
> After reviewing the logs, it looks like both reading and compactions stop on the bucket that is no longer allowing writes. My assumption is something has a lock on that bucket, usually this is caused by deletes. @zekena2 and @nward, are you doing deletes on the buckets that stop receiving writes? Also, @nward do you see the open file/socket issue that @zekena2 mentioned seeing?

We don't run any deletes, though the data retention runs of course. Would that qualify as "deletes"?
We see failures at all times of the day, do data retention deletes run constantly, or, only on set intervals (i.e. daily, or similar)?

I don't see a large number of open sockets, no. We restart influxd pretty fast (within a few minutes) after it fails, so perhaps we don't get to a point where it's a problem.

I also wonder if @zekena2's problem is a symptom rather than a cause - if the have lots of client requests (writes or reads) which are blocking, you might reasonably expect there to be lots of open sockets waiting for their requests to be processed.

**Comment by MaxOOOOON** [comment]
We have the same problem. The influx service freezes after 00:00UTC. Sometimes after 2 days, sometimes after a week.
In logs only: 
`lvl=debug msg=Request log_id=0giXx1PW000 service=http method=POST host=host path=/api/v2/write query="bucket=bucket" proto=HTTP/1.1 status_code=499 response_size=90 content_length=-1 referrer= remote=ip:port user_agent=Telegraf took=10019.286ms error="internal error" error_code="internal error"`
 and 400 errors on flux requests
` lvl=debug msg=Request log_id=0giXx1PW000 service=http method=POST host=IP:PORT path=/api/v2/query query="org=ORG" proto=HTTP/1.1 status_code=400 response_size=52 content_length=576 referrer= remote=REMOTE_host user_agent=influxdb-client-go took=11.008ms error=invalid error_code=invalid body="{\"dialect\":{\"annotations\":[\"datatype\",\"group\",\"default\"],\"delimiter\":\",\",\"header\":true},\"query\":\"QUERY\",\"type\":\"flux\"}"`

After freezing, the influx is not available for writing and after ~ 1 hour, when the RAM runs out, OOM killer comes and restarting influx service.



System info: Linux elka2023-influxdb 5.10.0-18-cloud-amd64 #1 SMP Debian 5.10.140-1 (2022-09-02) x86_64 GNU/Linux
Influx version: InfluxDB v2.4.0 (git: de247bab08) build_date: 2022-08-18T19:41:15Z
OS: Debian 11

**Comment by MaxOOOOON** [comment]
Yes, after errors in the logs, memory consumption begins to increase.  Initially, we had 6GB RAM and influx freezes for half an hour, then we added an additional 8GB and now influx frezzes for an hour.  during the day, 40-50% of the memory is used and when the nights are quiet(when the influx does not freeze and the grafana does not alert with all alerts), there are no memory jumps.

We do not have special deletions by timer.  use built-in bucket rotation and downsampling.

**Comment by MaxOOOOON** [comment]
By the way, when we use storage-compact-throughput-burst = 1331648, the influx seems to hang less often, on the 22nd I set this value and the influx has not yet hung up until today.  the default value was set earlier and we also tried to set storage-compact-throughput-burst = 8331648. with this config, in our case, influx hung more often.  perhaps this is a coincidence and next week the influx will hang again in a couple of days :(

**Comment by MaxOOOOON** [comment]
with parameters storage-max-concurrent-compactions = 1 and storage-compact-throughput-burst = 1331648 influx freeze after 4 days(It was April 7th). I also sent the log from April 7 to your email.

**Comment by jeffreyssmith2nd** [comment]
Hey @MaxOOOOON thanks for sending those logs over. Your issue has slightly different behavior than the issues above in that you are getting timeouts on the write rather than a full bucket lockout. Can you generate profiles when you're seeing those writes fail? You can watch the logs for `status_code=499` to know when the writes are timing out and profiles can be generated with
```
curl -o profiles.tar.gz "http://localhost:8086/debug/pprof/all?cpu=30s"
iostat -xd 1 30 > iostat.txt
```

**Comment by MaxOOOOON** [comment]
`curl -o profiles.tar.gz "http://localhost:8086/debug/pprof/all?cpu=30s"`
response 
`curl: (52) Empty reply from server`

I will try to execute the command `iostat -xd 1 30 > iostat.txt` during the problem, but I can send the logs of the atop command.

**Comment by mmarkus13** [comment]
Having the same issue. 

Did tweaking those parameters (`storage-compact-throughput-burst`  `storage-max-concurrent-compactions`) solve it for you, or have you done anything else since to fix it?

**Comment by MaxOOOOON** [comment]
I set the parameter storage-compact-throughput-burst = 665824 and the influx does not freeze, but now the cleaning does not work correctly. if look through the UI, there is no data in the bucket, but if you look directly at the server, the space does not decrease. After restarting the influxdb service, the space is freed up.

**Comment by mmarkus13** [comment]
Thanks for the info @zekena2 & @MaxOOOOON. Yeah I'd like to avoid unnecessary (scheduled) restarts, as it is only a temporary workaround, and creates gaps in data.

Is there any official info from @influxdata on this - possible RC?
Are they aware of the issue?

**Comment by john159753** [comment]
@MaxOOOOON - I'm glad I found your post, I have been struggling with the same issue since moving from v1.8, I thought I was alone. I am running on windows and was contemplating moving to a Linux install to see if it would help. It seems like at 00:00UTC something triggers, causes Influx to go into a bad state where it continually eats memory. During this bad state nothing is working, and influx responds to clients with "internal error". I even gave the server gobs and gobs of memory (like >100GB; hot-add because its a VM) to see if it was just a process that it needed to work through and it didn't seem like it was helping the situation. Every time I've needed to restart the service (or wait for it to crash on its own due to OOM). 

I'm going to try tweaking the same parameters too and see if it helps my situation.

**Comment by mmarkus13** [comment]
After endless of fine-tuning and testing we still experience this issue on regular basis - tho automatic restart when the issue happens at least mitigates the problem.

**Comment by orjanj** [comment]
Same issue here as well. Restarting is the way to go, temporarily fixes the issue until next time it appears.
Qualified guessing points to using the Flux language with large amounts of data, from my point of view (might also point to some compaction bugs?).
Might this fit with the same symptoms as others here?

Wondering if converting Flux-boards to InfluxQL can solve the issue on a longer term.

**Comment by zekena2** [comment]
I don't use flux and the queries are only in influxql but we still have this problem.

**Comment by AlbertasB** [comment]
Interestingly I've noticed this PR, https://github.com/influxdata/influxdb/pull/24584 its for 1.x but sounds oddly similar to what is happening with 2.x. Could this be the cause and proposed PR a solution? We also see these issues arrise at midnight, when new shards are being created.

**Comment by jeremow** [comment]
I have the same issue with InfluxDB 2.7.5. I have background scripts who delete some of the data in a cron and when I display in Grafana the data and I query the InfluxDB at the same time as it's deleting the data, the bucket is locked and I couldn't access it anymore until I restart influxd. I try to minimize the restarting by relaunching directly but still it can happen and display errors on our Grafana dashboards

**Comment by mmarkus13** [comment]
> Same issue with InfluxDB v2.7.10.
> 
> I don't run any delete scripts.
> 
> I can reproduce the issue with running heavy queries from Grafana, not every time, but after a few tries it usually locks up and spams this `"internal error"` to telegraf, when it is trying to write. Returning status code 499, with message `unexpected error writing points to database: engine: context canceled` or `unexpected error writing points to database: timeout`.
> 
> I have tried to tune the config every each way, but issue persists.


Actually we were able to reduce the almost daily restarts by distributing the telegraph instances so that not all data arrives in a bulk from a single source. That helped in our environment.

**Comment by devanbenz** [comment]
> Same issue with InfluxDB v2.7.10.
> 
> I don't run any delete scripts.
> 
> I can reproduce the issue with running heavy queries from Grafana, not every time, but after a few tries it usually locks up and spams this `"internal error"` to telegraf, when it is trying to write. Returning status code 499, with message `unexpected error writing points to database: engine: context canceled` or `unexpected error writing points to database: timeout`.
> 
> I have tried to tune the config every each way, but issue persists.

Would you be able to get profiles when the issue occurs and send them in this issue? If possible could you also grab the iostat https://unix.stackexchange.com/questions/225095/how-to-get-total-read-and-total-write-iops-in-linux IOPS and pair that with the profiles.

**Comment by hansole** [comment]
I have seen similar problems. I found it to be related to "resource demanding" queries while data being added at the same time. I have created a writer and reader in go that can trigger this deadlock. (I assume it is a deadlock.)

[https://github.com/hansole/influxdb2_deadlock](https://github.com/hansole/influxdb2_deadlock) (I was not able to upload a tar.gz file here, so I created a repository.)

Hopefully it should be possible to use this when investigating the cause of the deadlock, and that this is the same problem that I and other are observing.

**Comment by hansole** [comment]
I have been investigating some more, and found where it fails to obtain the lock when the bad state happens.

On the 2.7 branch, it fails to obtain the lock in these places:

[https://github.com/influxdata/influxdb/blob/68534d3ff2432298f5cc99288a62edfbf5065d13/tsdb/index/tsi1/log_file.go#L545](https://github.com/influxdata/influxdb/blob/68534d3ff2432298f5cc99288a62edfbf5065d13/tsdb/index/tsi1/log_file.go#L545)
[https://github.com/influxdata/influxdb/blob/68534d3ff2432298f5cc99288a62edfbf5065d13/tsdb/index/tsi1/log_file.go#L787](https://github.com/influxdata/influxdb/blob/68534d3ff2432298f5cc99288a62edfbf5065d13/tsdb/index/tsi1/log_file.go#L787)

You can see this if you use a debugger or simply put prints around them.

`gdb -ex=r --args ./bin/linux/influxd`

Run the reader and writer til they fails (timeout)

Interrupt the debugger and set breakpoints
```
(gdb) br tsdb/index/tsi1/log_file.go:545
Breakpoint 1 at 0x11e5709: file /home/hans/src/go/3rdParty/influxdb/tsdb/index/tsi1/log_file.go, line 545.
(gdb) br tsdb/index/tsi1/log_file.go:787
Breakpoint 2 at 0x11e7b3a: file /home/hans/src/go/3rdParty/influxdb/tsdb/index/tsi1/log_file.go, line 787.

gdb) cont
Continuing.
[Switching to Thread 0x7fff8d400640 (LWP 488649)]
```
Now run the reader program again in the other terminal. It will hit the break point, and you can see which file it tries to get a lock for.
```
Thread 13 "influxd" hit Breakpoint 2, github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*LogFile).MeasurementIterator (f=0xc0020fc780, ~r0=..., ~r0=...) at /home/hans/src/go/3rdParty/influxdb/tsdb/index/tsi1/log_file.go:787
787		f.mu.RLock()
(gdb) print f.path
$1 = "/home/hans/.influxdbv2/engine/data/8d4aaca62168ffd3/autogen/12/index/0/L0-00000006.tsl"
(gdb) n
788		defer f.mu.RUnlock()
(gdb) cont
Continuing.

Thread 13 "influxd" hit Breakpoint 2, github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*LogFile).MeasurementIterator (f=0xc001be10e0, ~r0=..., ~r0=...) at /home/hans/src/go/3rdParty/influxdb/tsdb/index/tsi1/log_file.go:787
787		f.mu.RLock()
(gdb) print f.path
$2 = "/home/hans/.influxdbv2/engine/data/8d4aaca62168ffd3/autogen/12/index/1/L0-00000006.tsl"
(gdb) n
788		defer f.mu.RUnlock()
(gdb) cont
Continuing.
```
This has to be done for several files. (5-7 more when I was testing)
```
Thread 11 "influxd" hit Breakpoint 2, github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*LogFile).MeasurementIterator (f=0xc0022b0c30, ~r0=..., ~r0=...) at /home/hans/src/go/3rdParty/influxdb/tsdb/index/tsi1/log_file.go:787
787		f.mu.RLock()
(gdb) print f.path
$7 = "/home/hans/.influxdbv2/engine/data/8d4aaca62168ffd3/autogen/12/index/6/L0-00000005.tsl"
(gdb) n
2025-05-04T20:16:17.479347Z	error	Unable to write gathered points	{"log_id": "0wJEhA80000", "service": "scraper", "scraper-name": "new target", "error": "timeout"}
788		defer f.mu.RUnlock()
(gdb) cont
Continuing.

Thread 11 "influxd" hit Breakpoint 2, github.com/influxdata/influxdb/v2/tsdb/index/tsi1.(*LogFile).MeasurementIterator (f=0xc0022b0b40, ~r0=..., ~r0=...) at /home/hans/src/go/3rdParty/influxdb/tsdb/index/tsi1/log_file.go:787
787		f.mu.RLock()
(gdb) print f.path
$8 = "/home/hans/.influxdbv2/engine/data/8d4aaca62168ffd3/autogen/12/index/7/L0-00000009.tsl"
(gdb) n
```
Lock request blocks. Someone is holding the lock. for /home/hans/.influxdbv2/engine/data/8d4aaca62168ffd3/autogen/12/index/7/L0-00000009.tsl.

It was a bit harder to get it reproduced for the writer in debugger, but adding prints before and after the Lock() function shows the problem.
```
	fmt.Println("DEBUG: AddSeriesList() locking", f.path)
	f.mu.Lock()
	fmt.Println("DEBUG: AddSeriesList()  locked", f.path)
```
```
	fmt.Println("DEBUG: MeasurementIterator() locking", f.path)
	f.mu.RLock()
	fmt.Println("DEBUG: MeasurementIterator()  locked", f.path)
```
```
DEBUG: AddSeriesList() locking /home/hans/.influxdbv2/engine/data/8d4aaca62168ffd3/autogen/12/index/3/L0-00000009.tsl
DEBUG: AddSeriesList()  locked /home/hans/.influxdbv2/engine/data/8d4aaca62168ffd3/autogen/12/index/3/L0-00000009.tsl
DEBUG: AddSeriesList() locking /home/hans/.influxdbv2/engine/data/8d4aaca62168ffd3/autogen/12/index/5/L0-00000006.tsl
DEBUG: AddSeriesList()  locked /home/hans/.influxdbv2/engine/data/8d4aaca62168ffd3/autogen/12/index/5/L0-00000006.tsl
DEBUG: AddSeriesList() locking /home/hans/.influxdbv2/engine/data/8d4aaca62168ffd3/autogen/12/index/1/L0-00000006.tsl
DEBUG: AddSeriesList()  locked /home/hans/.influxdbv2/engine/data/8d4aaca62168ffd3/autogen/12/index/1/L0-00000006.tsl
DEBUG: AddSeriesList() locking /home/hans/.influxdbv2/engine/data/8d4aaca62168ffd3/autogen/12/index/2/L0-00000006.tsl
```
```
DEBUG: MeasurementIterator() locking /home/hans/.influxdbv2/engine/data/8d4aaca62168ffd3/autogen/12/index/0/L0-00000006.tsl
DEBUG: MeasurementIterator()  locked /home/hans/.influxdbv2/engine/data/8d4aaca62168ffd3/autogen/12/index/0/L0-00000006.tsl
DEBUG: MeasurementIterator() locking /home/hans/.influxdbv2/engine/data/8d4aaca62168ffd3/autogen/12/index/1/L0-00000006.tsl
DEBUG: MeasurementIterator()  locked /home/hans/.influxdbv2/engine/data/8d4aaca62168ffd3/autogen/12/index/1/L0-00000006.tsl
DEBUG: MeasurementIterator() locking /home/hans/.influxdbv2/engine/data/8d4aaca62168ffd3/autogen/12/index/2/L0-00000006.tsl
```

I have not figured out what causes the lock to be held (who is not releasing it), but I hope this will be helpful input for your investigation.

Regards,

Hans

**Comment by hansole** [comment]
I have done some more investigation, and I might have found the actual problem, or at least part of it.
The TagValueIterator function for FileLog
https://github.com/influxdata/influxdb/blob/68534d3ff2432298f5cc99288a62edfbf5065d13/tsdb/index/tsi1/log_file.go#L419-L433
Calls TagValueIterator function for logTagKey
https://github.com/influxdata/influxdb/blob/68534d3ff2432298f5cc99288a62edfbf5065d13/tsdb/index/tsi1/log_file.go#L432
while holding the file lock in read mode.
This function
https://github.com/influxdata/influxdb/blob/68534d3ff2432298f5cc99288a62edfbf5065d13/tsdb/index/tsi1/log_file.go#L1387-L1396
then locks the same file (at least that is my understanding of how this code works, and it looks like that from debug printing) once again in read mode. This seems to be no problem when only reading. 

If a writer thread/goroutine tries to get read-write lock for the same lock between those two calls, the second attempt to (re-)lock the file hangs. It has to wait for the writer thread. The writer thread has to wait for us to release the read lock. We have a deadlock. (I'm quite new to Go, so my understanding of goroutines/threads and locks might be wrong.)

Since (*logTagKey) TagValueIterator is a public function it should continue to lock the file. So the solution seems to be to change in (*FileLog) TagValueIterator function to release the lock before calling.
```
func (f *LogFile) TagValueIterator(name, key []byte) TagValueIterator {
	f.mu.RLock()

	mm, ok := f.mms[string(name)]
	if !ok {
		f.mu.RUnlock()
		return nil
	}

	tk, ok := mm.tagSet[string(key)]
	if !ok {
		f.mu.RUnlock()
		return nil
	}
	f.mu.RUnlock()
	return tk.TagValueIterator()
}
```
The above change to the function seems to solve the deadlock problem for the test case posted here in this issue earlier. I have not checked if it works with our production setup, and if it might result in data corruption or other problems.

I hope someone with a better understanding of InfluxDB can have a look at it, and if it indeed solves the problem, provide a fix that can be released.

Regards,

Hans

**Comment by chrizzgapit** [comment]
It looks like a fix is coming: https://github.com/influxdata/influxdb/commit/2e842ff21d159cce6ed66266b35706846afe0fe8.
I am testing a build with that commit and using a modified version of the test programs by @hansole. The test has been running since Friday afternoon without triggering the problem.

**Comment by hansole** [comment]
Thanks for the information, @chrizzgapit. That is really good news! I was not aware that a fix was already underway due to no updates in this issue. Looks like I just re-discovered the same problem and solution, so it better fix the problem :)

**Comment by hansole** [comment]
Do anyone know when a new release with this fix can be expected?

**Comment by hansole** [comment]
Thanks @chrizzgapit. I just not good at tracking the activity

**Comment by hansole** [comment]
Seems to be no updated rpm or deb packages, or am I looking in the wrong places?

**Comment by john159753** [comment]
I think this actually fixed it. I was consistently getting the error every day around 00:00 UTC (sometimes in repetition).
Since applying 2.7.12, this behavior has stopped! Thank you all!

**Comment by DARSHAN-THE-CODER** [comment]
@john159753  can u confirm if you didn't face this issue again after upgrading to v2.7.12 ?

**Comment by raji-sridar** [comment]
Will InfluxDB v2.7.12 be marked latest stable version to get this fix in production servers?

**Comment by davidby-influx** [comment]
Yes, 2.7.12 is a stable, released version
